### PR TITLE
Added specs for Physical Server details

### PR DESCRIPTION
**Pull request by MaysaMacedo**
This PR contains specs for:
- GET physical server
- GET physical server with invalid ID
- GET physical server with invalid credentials
- REST Firmware details
- REST Hardware details
- REST NIC details 
- REST Network Adater port details
- REST Asset details
- REST VM details

### CODE REVIEW

File: spec/requests/physical_servers_spec.rb [hunk 1]
```diff
@@ -10,6 +10,155 @@
...
+        expect(response).to have_http_status(:forbidden)
+        expect(response.parsed_body["error"]).to include("kind" => "forbidden")
+      end
+    end
+
+    context "with valid id" do
+      it "returns both id and href" do
+        api_basic_authorize action_identifier(:physical_servers, :read, :resource_actions, :get)
+        ps = FactoryGirl.create(:physical_server)
+        ps2 = FactoryGirl.create(:physical_server)
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Is there a reason you're creating a second, unused physical server here?

File: spec/requests/physical_servers_spec.rb [hunk 2]
```diff
@@ -10,6 +10,155 @@
...
+
+        get api_physical_server_url(nil, ps.id)
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Nit, and not totally necessary, but you can omit the `.id` here and elsewhere

File: spec/requests/physical_servers_spec.rb [hunk 3]
```diff
@@ -10,6 +10,155 @@
...
+        get api_physical_server_url(nil, ps)
+
+        expect(response).to have_http_status(:forbidden)
+        expect(response.parsed_body["error"]).to include("kind" => "forbidden")
+      end
+    end
+
+    context "with valid id" do
+      it "returns both id and href" do
+        api_basic_authorize action_identifier(:physical_servers, :read, :resource_actions, :get)
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Prefer parentheses for these non-DSL method calls 😀

File: spec/requests/physical_servers_spec.rb [hunk 4]
```diff
@@ -10,6 +10,155 @@
...
+
+        get api_physical_server_url(nil, 999_999)
+
+        expect(response).to have_http_status(:not_found)
+      end
+    end
+
+    context "with hardware attribute" do
+      it "retrieve hardware details" do
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Given that this and the following tests are largely variations on the same theme (retrieving additional attributes from the physical server), and given that request specs are in general expensive, WDYT about combining them into one call where you ask for everything at once?

File: spec/requests/physical_servers_spec.rb [hunk 5]
```diff
@@ -10,6 +10,155 @@
         expect_single_resource_query("ems_ref" => "A59D5B36821111E1A9F5E41F13ED4F6A")
       end
     end
+
+    context "without an appropriate role" do
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Generally prefer context blocks when they are used to establish some shared setup through before blocks, lets, etc.. I think it's OK to put this "context" in test description itself

File: spec/requests/physical_servers_spec.rb [hunk 6]
```diff
@@ -10,6 +10,155 @@
...
+      it "fails to retrieve physical servers" do
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Nit, but the word "fails" is misleading. It succeeds in doing the right thing, which is to forbid access to the physical server

**Comment by MaysaMacedo** [comment]
@imtayadeway Thanks for your feedback. I've added the changes required.

File: spec/requests/physical_servers_spec.rb [hunk 7]
```diff
@@ -10,6 +10,85 @@
...
+        api_basic_authorize(action_identifier(:physical_servers, :read, :resource_actions, :get))
+
+        vm = FactoryGirl.create(:vm)
+        host = FactoryGirl.create(:host, :vms => [vm])
+
+        asset_details = FactoryGirl.create(:asset_details)
+
+        network = FactoryGirl.create(:network)
+        gd1 = FactoryGirl.create(:guest_device, :network => network)
+        gd1.update_attributes!(:device_type => 'ethernet')
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Can this be inlined, i.e.:

```ruby
gd1 = FactoryGirl.create(:guest_device, :network => network, :device_type => "ethernet")
```

File: spec/requests/physical_servers_spec.rb [hunk 8]
```diff
@@ -10,6 +10,85 @@
...
+            )
+          )
+        }
+        expect(response.parsed_body).to include(expected)
+        expect(response).to have_http_status(:ok)
+
+        expect(response.parsed_body['hardware']['ports'].count).to eq(1)
+        expect(response.parsed_body['hardware']['ports'].first['device_type']).to eq('ethernet')
+        expect(response.parsed_body['host']['vms'].count).to eq(1)
+        expect(response.parsed_body['host']['vms'].first['host_id']).to eq(host.id.to_s)
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: I think you can express all the above as a single expectation: 

```ruby
expected = {
  "host"          => a_hash_including(
    "physical_server_id" => ps.id.to_s,
    "vms" => [
      a_hash_including("host_id" => host.id.to_s)
    ]
  ),
  "asset_details" => a_hash_including("id" => asset_details.id.to_s),
  "hardware"      => a_hash_including(
    "id"        => hardware.id.to_s,
    "firmwares" => a_collection_including(
      a_hash_including("resource_id" => hardware.id.to_s)
    ),
    "nics"      => a_collection_including(
      a_hash_including("hardware_id" => hardware.id.to_s)
    ),
    "ports" => [
      a_hash_including("device_type" => "ethernet")
    ]
  )
}
expect(response.parsed_body).to include(expected)
```

File: spec/requests/physical_servers_spec.rb [hunk 9]
```diff
@@ -10,6 +10,85 @@
...
+
+        firmware = FactoryGirl.create(:firmware)
+        hardware = FactoryGirl.create(:hardware, :firmwares => [firmware], :guest_devices => [gd1, gd2])
+        network.update_attributes!(:hardware_id => hardware.id.to_s)
+
+        comp_system = FactoryGirl.create(:computer_system, :hardware => hardware)
+        ps = FactoryGirl.create(:physical_server, :computer_system => comp_system, :asset_details => asset_details, :host => host)
+
+        get api_physical_server_url(nil, ps), :params => {:attributes => "host,host.vms,asset_details,hardware,hardware.firmwares,hardware.nics,hardware.ports"}
+        expected = {
```
Comment (imtayadeway) [inline spec/requests/physical_servers_spec.rb]: Nit, but if you add a newline between this and the `get` above it helps clearly separate the execution and verification steps

**Comment by imtayadeway** [comment]
Thanks @MaysaMacedo ! 🥇

**Pull request merged**